File: core/types/money.ts
/**
 * Money + Cost. Micro-unit integer math; no floats. Mirrors the Python impl.
 */

export type Currency = string;

export class CurrencyError extends Error {}
export class AmountError extends Error {}

const MAX = (1n << 63n) - 1n;
const MIN = -(1n << 63n);

export class Money {
  readonly micro: bigint;
  readonly currency: Currency;

  constructor(micro: bigint | number | string, currency: Currency) {
    const m = typeof micro === "bigint" ? micro : BigInt(micro);
    if (m < MIN || m > MAX) throw new AmountError("micro out of range");
    if (!currency) throw new CurrencyError("currency required");
    this.micro = m;
    this.currency = currency.toUpperCase();
  }

  static zero(currency: Currency = "USDC"): Money {
    return new Money(0n, currency);
  }

  static fromMajor(major: number | string, currency: Currency = "USDC"): Money {
    // Convert with 6-decimal precision using bigint math.
    const [whole, frac = ""] = String(major).split(".");
    const fracPadded = (frac + "000000").slice(0, 6);
    const micro = BigInt(whole) * 1_000_000n + BigInt(fracPadded || "0") * BigInt(frac.length < 6 ? 1 : 1);
    return new Money(micro, currency);
  }

  toMajor(): string {
    const sign = this.micro < 0n ? "-" : "";
    const abs = this.micro < 0n ? -this.micro : this.micro;
    const whole = abs / 1_000_000n;
    const frac = (abs % 1_000_000n).toString().padStart(6, "0");
    return `${sign}${whole.toString()}.${frac}`;
  }

  toString(): string {
    return `${this.toMajor()} ${this.currency}`;
  }

  private assertSame(o: Money): void {
    if (this.currency !== o.currency)
      throw new CurrencyError(`currency mismatch: ${this.currency} vs ${o.currency}`);
  }

  add(o: Money): Money {
    this.assertSame(o);
    return new Money(this.micro + o.micro, this.currency);
  }
  sub(o: Money): Money {
    this.assertSame(o);
    return new Money(this.micro - o.micro, this.currency);
  }
  neg(): Money {
    return new Money(-this.micro, this.currency);
  }
  mul(n: number | bigint): Money {
    const factor = typeof n === "bigint" ? n : BigInt(n);
    return new Money(this.micro * factor, this.currency);
  }

  lt(o: Money): boolean {
    this.assertSame(o);
    return this.micro < o.micro;
  }
  le(o: Money): boolean {
    this.assertSame(o);
    return this.micro <= o.micro;
  }
  gt(o: Money): boolean {
    this.assertSame(o);
    return this.micro > o.micro;
  }
  ge(o: Money): boolean {
    this.assertSame(o);
    return this.micro >= o.micro;
  }
  isZero(): boolean {
    return this.micro === 0n;
  }

  equals(o: Money): boolean {
    return this.micro === o.micro && this.currency === o.currency;
  }
}

export type CostKind =
  | "cpu_ms"
  | "gpu_ms"
  | "net_bytes"
  | "disk_bytes"
  | "api_call"
  | "tool_ms";

export class Cost {
  constructor(public ru: bigint, public kind: CostKind) {
    if (ru < 0n) throw new AmountError("ru must be non-negative");
  }
  add(o: Cost): Cost {
    if (o.kind !== this.kind)
      throw new AmountError(`cannot add costs: ${this.kind} + ${o.kind}`);
    return new Cost(this.ru + o.ru, this.kind);
  }
  mul(n: number | bigint): Cost {
    const f = typeof n === "bigint" ? n : BigInt(n);
    return new Cost(this.ru * f, this.kind);
  }
}

File: core/types/money.py
"""
Money and cost representation.

All monetary values are stored as integer micro-units (1e-6 of the base unit)
to avoid floating-point errors. Currency is an explicit string. Conversion
rates are loaded from the FX oracle and cached; rates are versioned.

Costs and budgets are also expressed in micro-units, in the Automaton's
configured base currency (default: USDC).
"""
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal
from typing import Final

MAX_MICRO: Final[int] = 2**63 - 1
MIN_MICRO: Final[int] = -(2**63)


class CurrencyError(ValueError):
    pass


class AmountError(ValueError):
    pass


@dataclass(frozen=True, slots=True)
class Money:
    """Immutable money amount in micro-units of `currency`."""

    micro: int
    currency: str

    def __post_init__(self) -> None:
        if not isinstance(self.micro, int):
            raise AmountError(f"micro must be int, got {type(self.micro).__name__}")
        if not (MIN_MICRO <= self.micro <= MAX_MICRO):
            raise AmountError("micro out of range")
        if not isinstance(self.currency, str) or not self.currency:
            raise CurrencyError("currency must be a non-empty string")
        object.__setattr__(self, "currency", self.currency.upper())

    # -- constructors -------------------------------------------------
    @classmethod
    def zero(cls, currency: str = "USDC") -> "Money":
        return cls(0, currency)

    @classmethod
    def from_major(cls, major: Decimal | int | float | str, currency: str = "USDC") -> "Money":
        d = Decimal(str(major))
        if d != d.to_integral_value() and -d.as_tuple().exponent > 6:
            raise AmountError("more than 6 decimal places")
        return cls(int(d * Decimal(1_000_000)), currency)

    # -- conversions --------------------------------------------------
    def to_major(self) -> Decimal:
        return Decimal(self.micro) / Decimal(1_000_000)

    def __str__(self) -> str:
        return f"{self.to_major():.6f} {self.currency}"

    # -- arithmetic (same currency) -----------------------------------
    def _assert_same(self, other: "Money") -> None:
        if self.currency != other.currency:
            raise CurrencyError(
                f"currency mismatch: {self.currency} vs {other.currency}"
            )

    def __add__(self, other: "Money") -> "Money":
        self._assert_same(other)
        return Money(self.micro + other.micro, self.currency)

    def __sub__(self, other: "Money") -> "Money":
        self._assert_same(other)
        return Money(self.micro - other.micro, self.currency)

    def __neg__(self) -> "Money":
        return Money(-self.micro, self.currency)

    def __mul__(self, factor: int) -> "Money":
        if not isinstance(factor, int):
            raise AmountError("can only multiply Money by int")
        return Money(self.micro * factor, self.currency)

    __rmul__ = __mul__

    # -- comparisons --------------------------------------------------
    def __lt__(self, other: "Money") -> bool:
        self._assert_same(other)
        return self.micro < other.micro

    def __le__(self, other: "Money") -> bool:
        self._assert_same(other)
        return self.micro <= other.micro

    def __gt__(self, other: "Money") -> bool:
        self._assert_same(other)
        return self.micro > other.micro

    def __ge__(self, other: "Money") -> bool:
        self._assert_same(other)
        return self.micro >= other.micro

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Money):
            return NotImplemented
        return self.micro == other.micro and self.currency == other.currency

    def __hash__(self) -> int:
        return hash((self.micro, self.currency))


@dataclass(frozen=True, slots=True)
class Cost:
    """A non-monetary cost expressed in abstract resource units (RU)."""

    ru: int
    kind: str  # 'cpu_ms' | 'gpu_ms' | 'net_bytes' | 'disk_bytes' | 'api_call' | 'tool_ms'

    def __post_init__(self) -> None:
        if not isinstance(self.ru, int) or self.ru < 0:
            raise AmountError("ru must be a non-negative int")
        if not isinstance(self.kind, str) or not self.kind:
            raise AmountError("kind must be a non-empty string")

    def __add__(self, other: "Cost") -> "Cost":
        if self.kind != other.kind:
            raise AmountError(f"cannot add costs of different kinds: {self.kind} + {other.kind}")
        return Cost(self.ru + other.ru, self.kind)

    def __mul__(self, factor: int) -> "Cost":
        return Cost(self.ru * factor, self.kind)
